### ISSUE TITLE
linearsolverloader pardiso interface out of date

### ISSUE THREAD

**Issue by svigerske** [coin-or/Ipopt#153]
Issue created by migration from Trac.

Original creator: jdh

Original creation time: 2011-03-17 13:44:22

Assignee: ipopt-team

Version: 3.9

PARDISO has a new interface (as of v4.0.0?) that contains additional arguments to the old interface.

The HAVE_PARDISO_NEWINTERFACE macro is used to establish this existence at compile time if --have-pardiso= is used at configure. Otherwise the old interface is assumed.

Further, the linear solver library loader interface still uses the old version.

Looking at Ipopt/configure.ac it seems that the new interface can be differentiated from the old one as defines the symbol pardiso_ipopt_newinterface. This could be exploited by the linear solver loader to use the correct interface.

I suspect that the new interface should be made the default in IpPardisoSolverInterface and the logic inverted so that ONLY if the old interface is detected at configure time does IpPardisoSolverInterface use the old interface. This then allows linear solver loader to detect the old interface at runtime and just ignore the additional arguments (note: I have not checked if this can be done safely).

Jonathan.

**Comment by svigerske** [comment]
Comment by jdh created at 2011-04-07 20:29:41

I've now got a prototype version of linearsolverloader that copes with this by detecting the presence of the new interface and providing wrappers to it for the old interface. I remain convinced that we really should do this the other way around for future proofing reasons (presumably at some stage IPOPT will support the iterative method part of PARDISO). I'm still waiting for the legal dept to greenlight code contribution to IPOPT, so can't post it here yet.

This has highlighted another problem however - at present if no --with-pardiso option is used then regardless of OMP_NUM_THREADS the pardiso number of threads is set to 1 as HAVE_PARDISO_PARALLEL is not set. This leads to the confusing situation of no speedup up on wall clock time, but an n-fold increase in CPU time where n is the number of CPUs. Again, this is probably best dealt with by defaulting to assuming a parallel Pardiso unless a non-parallel version is supplied. A simple boolean function could be added to linearsolverloader to say yes or no to the parallel version being present.

Jonathan.

**Comment by svigerske** [comment]
Comment by @svigerske created at 2011-04-26 12:35:05

How is the status with the legal dept?

I agree that per default one could assume the latest Pardiso release, and support older versions only for a limited time.

If you need help on changing the linear solver loader, let me know.

Stefan

**Comment by svigerske** [comment]
Attachment [PardisoLoader.c](https://gist.githubusercontent.com/svigerske/64d55c1ba42b8443337c1fda881ce3b3/raw/42a4ea83877681d1467e6376572be691f5e8203f/Ipopt%20issue%20153%20attachment%20PardisoLoader.c) by jdh created at 2011-06-02 14:46:05

**Comment by svigerske** [comment]
Comment by jdh created at 2011-06-02 14:51:36

Looking at the new PARDISO licencing regieme I'm not convinced that developing a patch for IPOPT qualifies for the free licence. As a result I've not been able to polish the patch I just attached (especially in terms of rearranging the logic) in the ways I'd like to.

However, that attachment was sufficient to prevent crashes when calling the new interface PARDISO using the old-style code in IpPardisoSolverInterface.cpp

To tidy this up what probably needs doing:
* Invert logic so that new style call is the default, change linearsolverloader wrapper to have a wrap_old function rather than a wrap_new function.
* Fix default number of threads in IpPardisoSolverInterface.cpp to be non-serial by default - otherwise users are likely to get confused over why they're not getting any parallel speedup.

Apologies I can't deliver something more concrete.

Jonathan.

**Comment by svigerske** [comment]
Comment by @svigerske created at 2011-06-14 13:58:08

Thanks, I took the liberty to modify the code in Ipopt/trunk.

I did the point with assuming a new pardiso lib by default and have a wrapper in case its an old one.

Regarding the number of threads, if the linear solver loader is used, then IpPardisoSolverInterface assumes a parallel version and checks for the OMP_NUM_THREADS variable, but it does not give an error if this variable is not set.
The linear solver loader recognizes whether it is a serial or parallel version of the pardiso library and sets the number of threads parameter to 1 if it is a serial version.

Stefan

PS: Somehow, Pardiso seem to run with 2 cores even if the number of threads is specified as 1 in IPARM, but OMP_NUM_THREADS is not set. Setting OMP_NUM_THREADS explicitly to 1 restricts it to run with one thread only.
